canton/community/ledger/ledger-api-core/src/main/scala/com/digitalasset/canton/platform/apiserver/services/admin/ApiPackageManagementService.scala

```diff
@@ -131,57 +133,138 @@ private[apiserver] final class ApiPackageManagementService private (
+                      s"Typechecking upgrades for $pkgId2 failed with unknown error."
                     ),
                   )
                   .asGrpcError
               )
           }
-        } yield ()
-      case None => {
-        logger.info(s"Package $upgradingPackageId does not upgrade anything.")
-        Future(())
       }
     }
   }
 
+  def typecheckUpgrades(typecheckPhase: TypecheckUpgrades.UploadPhaseCheck, optDar1: Option[(Ref.PackageId, Ast.Package)], optDar2: Option[(Ref.PackageId, Ast.Package)])(implicit
+                                                                                                                      loggingContext: LoggingContextWithTrace
+  ): Future[Unit] = {
+    (optDar1, optDar2) match {
+      case (None, _) | (_, None) =>
+        Future.unit
+
+      case (Some((pkgId1, pkg1)), Some((pkgId2, pkg2))) =>
+        strictTypecheckUpgrades(typecheckPhase, Some((pkgId1, pkg1)), pkgId2, Some(pkg2))
+    }
+  }
+
+  private def validateUpgrade(upgradingDar: Dar[(Ref.PackageId, Ast.Package)])(implicit
+      loggingContext: LoggingContextWithTrace
+  ): Future[Unit] = {
+    val (upgradingPackageId, upgradingPackage) = upgradingDar.main
+    val optUpgradingDar = Some((upgradingPackageId, upgradingPackage))
+    logger.info(s"Uploading DAR file for $upgradingPackageId in submission ID ${loggingContext.serializeFiltered("submissionId")}.")
+    existingVersionedPackageId(upgradingDar.main) match {
+      case Some(uploadedPackageId) =>
+        if (uploadedPackageId == upgradingPackageId) {
+          logger.info(s"Ignoring upload of package $upgradingPackageId as it has been previously uploaded")
+          Future.unit
+        } else {
+          logger.info(s"Package $upgradingPackageId can not be uploaded - $uploadedPackageId is already uploaded and at version ${upgradingPackage.metadata.version}")
+          Future.failed(Validation.UpgradeVersion.Error(uploadedPackageId, upgradingPackageId, upgradingPackage.metadata.version).asGrpcError)
+        }
+
+      case None =>
+        upgradingPackage.metadata.upgradedPackageId match {
+          case Some(upgradedPackageId) =>
+            for {
+              optUpgradedDar <- lookupDar(upgradedPackageId)
+              _ <- strictTypecheckUpgrades(TypecheckUpgrades.DarCheck, optUpgradingDar, upgradedPackageId, optUpgradedDar.map(_._2))
+              optMaximalDar <- maximalVersionedDar(upgradedPackageId, upgradingDar.main)
+              _ <- typecheckUpgrades(TypecheckUpgrades.MaximalDarCheck, optMaximalDar, optUpgradingDar)
+              optMinimalDar <- minimalVersionedDar(upgradedPackageId, upgradingDar.main)
+              _ <- typecheckUpgrades(TypecheckUpgrades.MinimalDarCheck, optUpgradingDar, optMinimalDar)
+              _ = logger.info(s"Typechecking upgrades for $upgradingPackageId succeeded.")
+            } yield ()
+
+          case None =>
```

Comment:
No check if `upgradedPackageId` is not populated in the pgk meta? Maybe I am not following precisely, but if we upload an upgrading-incompatible package that can brick the participant (as the insertion in `PackageMetadata` might fail indefinitely). Feel free to ignore if you are planning to address in a subsequent PR.

Reply:
Thanks, I'd forgotten to add this case - now done!